File: src/css/rules/container.zig
pub const css = @import("../css_parser.zig");
const Result = css.Result;
const Printer = css.Printer;
const PrintErr = css.PrintErr;
const Location = css.css_rules.Location;
const CustomIdent = css.css_values.ident.CustomIdent;
const CustomIdentFns = css.css_values.ident.CustomIdentFns;
const QueryFeature = css.media_query.QueryFeature;
const QueryConditionFlags = css.media_query.QueryConditionFlags;
const Operator = css.media_query.Operator;

pub const ContainerName = struct {
    v: css.css_values.ident.CustomIdent,
    pub fn parse(input: *css.Parser) Result(ContainerName) {
        const ident = switch (CustomIdentFns.parse(input)) {
            .result => |vv| vv,
            .err => |e| return .{ .err = e },
        };

        // todo_stuff.match_ignore_ascii_case;
        if (bun.strings.eqlCaseInsensitiveASCIIICheckLength("none", ident.v) or
            bun.strings.eqlCaseInsensitiveASCIIICheckLength("and", ident.v) or
            bun.strings.eqlCaseInsensitiveASCIIICheckLength("not", ident.v) or
            bun.strings.eqlCaseInsensitiveASCIIICheckLength("or", ident.v))
            return .{ .err = input.newUnexpectedTokenError(.{ .ident = ident.v }) };

        return .{ .result = ContainerName{ .v = ident } };
    }

    const This = @This();

    pub fn toCss(this: *const This, dest: *Printer) PrintErr!void {
        return try CustomIdentFns.toCss(&this.v, dest);
    }

    pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
        return css.implementDeepClone(@This(), this, allocator);
    }
};

pub const ContainerNameFns = ContainerName;
pub const ContainerSizeFeature = QueryFeature(ContainerSizeFeatureId);

pub const ContainerSizeFeatureId = enum {
    /// The [width](https://w3c.github.io/csswg-drafts/css-contain-3/#width) size container feature.
    width,
    /// The [height](https://w3c.github.io/csswg-drafts/css-contain-3/#height) size container feature.
    height,
    /// The [inline-size](https://w3c.github.io/csswg-drafts/css-contain-3/#inline-size) size container feature.
    @"inline-size",
    /// The [block-size](https://w3c.github.io/csswg-drafts/css-contain-3/#block-size) size container feature.
    @"block-size",
    /// The [aspect-ratio](https://w3c.github.io/csswg-drafts/css-contain-3/#aspect-ratio) size container feature.
    @"aspect-ratio",
    /// The [orientation](https://w3c.github.io/csswg-drafts/css-contain-3/#orientation) size container feature.
    orientation,

    pub const valueType = css.DeriveValueType(@This(), ValueTypeMap).valueType;

    pub const ValueTypeMap = .{
        .width = css.MediaFeatureType.length,
        .height = css.MediaFeatureType.length,
        .@"inline-size" = css.MediaFeatureType.length,
        .@"block-size" = css.MediaFeatureType.length,
        .@"aspect-ratio" = css.MediaFeatureType.ratio,
        .orientation = css.MediaFeatureType.ident,
    };

    pub fn asStr(this: *const @This()) []const u8 {
        return css.enum_property_util.asStr(@This(), this);
    }

    pub fn parse(input: *css.Parser) Result(@This()) {
        return css.enum_property_util.parse(@This(), input);
    }

    pub fn toCss(this: *const @This(), dest: *Printer) PrintErr!void {
        return css.enum_property_util.toCss(@This(), this, dest);
    }

    pub fn toCssWithPrefix(this: *const @This(), prefix: []const u8, dest: *Printer) PrintErr!void {
        try dest.writeStr(prefix);
        try this.toCss(dest);
    }
};

/// Represents a style query within a container condition.
pub const StyleQuery = union(enum) {
    /// A style feature, implicitly parenthesized.
    feature: css.Property,

    /// A negation of a condition.
    not: *StyleQuery,

    /// A set of joint operations.
    operation: struct {
        /// The operator for the conditions.
        operator: css.media_query.Operator,
        /// The conditions for the operator.
        conditions: ArrayList(StyleQuery),

        pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
            return css.implementDeepClone(@This(), this, allocator);
        }
    },

    pub fn toCss(this: *const StyleQuery, dest: *Printer) PrintErr!void {
        switch (this.*) {
            .feature => |f| try f.toCss(dest, false),
            .not => |c| {
                try dest.writeStr("not ");
                return try css.media_query.toCssWithParensIfNeeded(
                    c,
                    dest,
                    c.needsParens(null, &dest.targets),
                );
            },
            .operation => |op| return css.media_query.operationToCss(
                StyleQuery,
                op.operator,
                &op.conditions,
                dest,
            ),
        }
    }

    pub fn parseFeature(input: *css.Parser) Result(StyleQuery) {
        const property_id = switch (css.PropertyId.parse(input)) {
            .result => |vv| vv,
            .err => |e| return .{ .err = e },
        };
        if (input.expectColon().asErr()) |e| return .{ .err = e };
        input.skipWhitespace();
        const opts = css.ParserOptions.default(input.allocator(), null);
        const feature: StyleQuery = .{
            .feature = switch (css.Property.parse(
                property_id,
                input,
                &opts,
            )) {
                .result => |vv| vv,
                .err => |e| return .{ .err = e },
            },
        };
        _ = input.tryParse(css.parseImportant, .{});
        return .{ .result = feature };
    }

    pub fn createNegation(condition: *StyleQuery) StyleQuery {
        return .{ .not = condition };
    }

    pub fn createOperation(operator: Operator, conditions: ArrayList(StyleQuery)) StyleQuery {
        return .{
            .operation = .{
                .operator = operator,
                .conditions = conditions,
            },
        };
    }

    pub fn needsParens(
        this: *const StyleQuery,
        parent_operator: ?Operator,
        _: *const css.Targets,
    ) bool {
        return switch (this.*) {
            .not => true,
            .operation => |op| op.operator == parent_operator,
            .feature => true,
        };
    }

    pub fn parseStyleQuery(input: *css.Parser) Result(@This()) {
        return .{ .err = input.newErrorForNextToken() };
    }

    pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
        return css.implementDeepClone(@This(), this, allocator);
    }
};

pub const ContainerCondition = union(enum) {
    /// A size container feature, implicitly parenthesized.
    feature: ContainerSizeFeature,
    /// A negation of a condition.
    not: *ContainerCondition,
    /// A set of joint operations.
    operation: struct {
        /// The operator for the conditions.
        operator: css.media_query.Operator,
        /// The conditions for the operator.
        conditions: ArrayList(ContainerCondition),

        pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
            return css.implementDeepClone(@This(), this, allocator);
        }
    },
    /// A style query.
    style: StyleQuery,

    const This = @This();

    pub fn parse(input: *css.Parser) Result(ContainerCondition) {
        return css.media_query.parseQueryCondition(
            ContainerCondition,
            input,
            QueryConditionFlags{
                .allow_or = true,
                .allow_style = true,
            },
        );
    }

    pub fn toCss(this: *const This, dest: *Printer) PrintErr!void {
        switch (this.*) {
            .feature => |f| try f.toCss(dest),
            .not => |c| {
                try dest.writeStr("not ");
                return try css.media_query.toCssWithParensIfNeeded(
                    c,
                    dest,
                    c.needsParens(null, &dest.targets),
                );
            },
            .operation => |op| try css.media_query.operationToCss(ContainerCondition, op.operator, &op.conditions, dest),
            .style => |query| {
                try dest.writeStr("style(");
                try query.toCss(dest);
                try dest.writeChar(')');
            },
        }
    }

    pub fn parseFeature(input: *css.Parser) Result(ContainerCondition) {
        const feature = switch (QueryFeature(ContainerSizeFeatureId).parse(input)) {
            .result => |vv| vv,
            .err => |e| return .{ .err = e },
        };
        return .{ .result = .{ .feature = feature } };
    }

    pub fn createNegation(condition: *ContainerCondition) ContainerCondition {
        return .{ .not = condition };
    }

    pub fn createOperation(operator: Operator, conditions: ArrayList(ContainerCondition)) ContainerCondition {
        return .{
            .operation = .{
                .operator = operator,
                .conditions = conditions,
            },
        };
    }

    pub fn parseStyleQuery(input: *css.Parser) Result(ContainerCondition) {
        const Fns = struct {
            pub inline fn adaptedParseQueryCondition(i: *css.Parser, flags: QueryConditionFlags) Result(StyleQuery) {
                return css.media_query.parseQueryCondition(StyleQuery, i, flags);
            }

            pub fn parseNestedBlockFn(_: void, i: *css.Parser) Result(ContainerCondition) {
                if (i.tryParse(
                    @This().adaptedParseQueryCondition,
                    .{
                        QueryConditionFlags{ .allow_or = true },
                    },
                ).asValue()) |res| {
                    return .{ .result = .{ .style = res } };
                }

                return .{ .result = .{
                    .style = switch (StyleQuery.parseFeature(i)) {
                        .result => |vv| vv,
                        .err => |e| return .{ .err = e },
                    },
                } };
            }
        };
        return input.parseNestedBlock(ContainerCondition, {}, Fns.parseNestedBlockFn);
    }

    pub fn needsParens(
        this: *const ContainerCondition,
        parent_operator: ?Operator,
        targets: *const css.Targets,
    ) bool {
        return switch (this.*) {
            .not => true,
            .operation => |op| op.operator == parent_operator,
            .feature => |f| f.needsParens(parent_operator, targets),
            .style => false,
        };
    }

    pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
        return css.implementDeepClone(@This(), this, allocator);
    }
};

/// A [@container](https://drafts.csswg.org/css-contain-3/#container-rule) rule.
pub fn ContainerRule(comptime R: type) type {
    return struct {
        /// The name of the container.
        name: ?ContainerName,
        /// The container condition.
        condition: ContainerCondition,
        /// The rules within the `@container` rule.
        rules: css.CssRuleList(R),
        /// The location of the rule in the source file.
        loc: Location,

        const This = @This();

        pub fn toCss(this: *const This, dest: *Printer) PrintErr!void {
            // #[cfg(feature = "sourcemap")]
            // dest.add_mapping(self.loc);

            try dest.writeStr("@container ");
            if (this.name) |*name| {
                try name.toCss(dest);
                try dest.writeChar(' ');
            }

            // Don't downlevel range syntax in container queries.
            const exclude = dest.targets.exclude;
            bun.bits.insert(css.targets.Features, &dest.targets.exclude, .media_queries);
            try this.condition.toCss(dest);
            dest.targets.exclude = exclude;

            try dest.whitespace();
            try dest.writeChar('{');
            dest.indent();
            try dest.newline();
            try this.rules.toCss(dest);
            dest.dedent();
            try dest.newline();
            try dest.writeChar('}');
        }

        pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) This {
            return css.implementDeepClone(@This(), this, allocator);
        }
    };
}

const bun = @import("bun");

const std = @import("std");
const ArrayList = std.ArrayListUnmanaged;

File: src/css/rules/container.rs
use crate as css;
use crate::css_rules::{CssRuleList, Location};
use crate::css_values::ident::CustomIdent;
use crate::media_query::{self, MediaFeatureType, Operator, QueryCondition, QueryFeature, ToCss};
use crate::properties::Property;
use crate::{PrintErr, Printer};

/// A [`<container-name>`](https://drafts.csswg.org/css-contain-3/#typedef-container-name).
pub struct ContainerName {
    pub v: CustomIdent,
}

impl ContainerName {
    pub fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        super::custom_ident_to_css(&self.v, dest)
    }
}

impl ContainerName {
    #[inline]
    pub fn deep_clone(&self, bump: &bun_alloc::Arena) -> Self {
        // PORT NOTE: `css.implementDeepClone` field-walk — `CustomIdent`
        // identity-copy (arena-owned slice pointer).
        Self {
            v: self.v.deep_clone(bump),
        }
    }
}

// ─── ContainerName parse ──────────────────────────────────────────────────
impl ContainerName {
    pub fn parse(input: &mut css::Parser) -> css::Result<ContainerName> {
        use crate::css_values::ident::CustomIdentFns;
        use bun_core::strings;
        let ident = match CustomIdentFns::parse(input) {
            Ok(vv) => vv,
            Err(e) => return Err(e),
        };

        // SAFETY: CustomIdent.v points into the parser source/arena (Phase A
        // lifetime erasure — see PORTING.md §AST crates).
        let v: &'static [u8] = unsafe { crate::arena_str(ident.v) };
        // todo_stuff.match_ignore_ascii_case;
        if strings::eql_any_case_insensitive_ascii(v, &[b"none", b"and", b"not", b"or"]) {
            return Err(input.new_unexpected_token_error(css::Token::Ident(v)));
        }

        Ok(ContainerName { v: ident })
    }
}

pub use ContainerName as ContainerNameFns;
pub type ContainerSizeFeature = QueryFeature<ContainerSizeFeatureId>;

#[derive(Clone, Copy, PartialEq, Eq, css::DefineEnumProperty)]
pub enum ContainerSizeFeatureId {
    /// The [width](https://w3c.github.io/csswg-drafts/css-contain-3/#width) size container feature.
    Width,
    /// The [height](https://w3c.github.io/csswg-drafts/css-contain-3/#height) size container feature.
    Height,
    /// The [inline-size](https://w3c.github.io/csswg-drafts/css-contain-3/#inline-size) size container feature.
    InlineSize,
    /// The [block-size](https://w3c.github.io/csswg-drafts/css-contain-3/#block-size) size container feature.
    BlockSize,
    /// The [aspect-ratio](https://w3c.github.io/csswg-drafts/css-contain-3/#aspect-ratio) size container feature.
    AspectRatio,
    /// The [orientation](https://w3c.github.io/csswg-drafts/css-contain-3/#orientation) size container feature.
    Orientation,
}

// `QueryFeature<FeatureId>` requires `FeatureId: FeatureIdTrait` at the type
// level, so this impl must be present for `ContainerSizeFeature` to resolve.
// `value_type` inlines the Zig `DeriveValueType` reflection; `to_css`/`from_str`
// delegate to `enum_property_util` (driven by the `EnumProperty` derive).
impl crate::media_query::FeatureIdTrait for ContainerSizeFeatureId {
    // Zig: pub const valueType = css.DeriveValueType(@This(), ValueTypeMap).valueType;
    // PORT NOTE: DeriveValueType is comptime reflection over ValueTypeMap; expanded inline.
    fn value_type(&self) -> MediaFeatureType {
        match self {
            Self::Width => MediaFeatureType::Length,
            Self::Height => MediaFeatureType::Length,
            Self::InlineSize => MediaFeatureType::Length,
            Self::BlockSize => MediaFeatureType::Length,
            Self::AspectRatio => MediaFeatureType::Ratio,
            Self::Orientation => MediaFeatureType::Ident,
        }
    }

    fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        css::enum_property_util::to_css(self, dest)
    }

    fn from_str(s: &[u8]) -> Option<Self> {
        <Self as css::EnumProperty>::from_ascii_case_insensitive(s)
    }
}

impl ContainerSizeFeatureId {
    pub fn to_css_with_prefix(
        &self,
        prefix: &[u8],
        dest: &mut Printer,
    ) -> core::result::Result<(), PrintErr> {
        dest.write_str(prefix)?;
        self.to_css(dest)
    }
}

/// Represents a style query within a container condition.
pub enum StyleQuery {
    /// A style feature, implicitly parenthesized.
    Feature(Property),

    /// A negation of a condition.
    Not(Box<StyleQuery>),

    /// A set of joint operations.
    Operation {
        /// The operator for the conditions.
        operator: Operator,
        /// The conditions for the operator.
        // PERF(port): was ArrayListUnmanaged fed input.arena() (parser arena);
        // Phase B decides bun_alloc::ArenaVec<'bump, _> vs global Vec crate-wide.
        conditions: Vec<StyleQuery>,
    },
}

impl ToCss for StyleQuery {
    fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        self.condition_to_css(dest)
    }
}

impl QueryCondition for StyleQuery {
    type Feature = Property;

    fn as_feature(&self) -> Option<&Property> {
        if let Self::Feature(f) = self {
            Some(f)
        } else {
            None
        }
    }
    fn as_not(&self) -> Option<&Self> {
        if let Self::Not(c) = self {
            Some(c)
        } else {
            None
        }
    }
    fn as_operation(&self) -> Option<(Operator, &[Self])> {
        if let Self::Operation {
            operator,
            conditions,
        } = self
        {
            Some((*operator, conditions))
        } else {
            None
        }
    }
    fn feature_to_css(f: &Property, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        f.to_css(dest, false)
    }

    fn parse_feature(input: &mut css::Parser) -> css::Result<Self> {
        let property_id = crate::properties::PropertyId::parse(input)?;
        input.expect_colon()?;
        input.skip_whitespace();
        // PORT NOTE: Zig threaded `(input.arena(), null)` here; Phase B
        // re-threads `&Bump` once `ParserOptions` carries the arena.
        let opts = css::ParserOptions::default(None);
        let feature = StyleQuery::Feature(Property::parse(property_id, input, &opts)?);
        let _ = input.try_parse(css::css_parser::parse_important);
        Ok(feature)
    }
    fn create_negation(condition: Box<Self>) -> Self {
        StyleQuery::Not(condition)
    }
    fn create_operation(operator: Operator, conditions: Vec<Self>) -> Self {
        StyleQuery::Operation {
            operator,
            conditions,
        }
    }
    fn parse_style_query(input: &mut css::Parser) -> css::Result<Self> {
        // Zig: `return .{ .err = input.newErrorForNextToken() }`
        Err(input.new_error_for_next_token())
    }
    fn needs_parens(&self, parent_operator: Option<Operator>, _targets: &css::Targets) -> bool {
        match self {
            StyleQuery::Not(_) => true,
            StyleQuery::Operation { operator, .. } => Some(*operator) == parent_operator,
            StyleQuery::Feature(_) => true,
        }
    }
}

impl StyleQuery {
    pub fn deep_clone(&self, bump: &bun_alloc::Arena) -> Self {
        // PORT NOTE: `css.implementDeepClone` variant-walk. `Operator` is `Copy`;
        // `Property` routes through `dc::property` until the per-variant
        // `DeepClone` derives land in `properties_generated.rs`.
        match self {
            Self::Feature(p) => Self::Feature(super::dc::property(p, bump)),
            Self::Not(c) => Self::Not(Box::new(c.deep_clone(bump))),
            Self::Operation {
                operator,
                conditions,
            } => Self::Operation {
                operator: *operator,
                conditions: conditions.iter().map(|c| c.deep_clone(bump)).collect(),
            },
        }
    }
}

pub enum ContainerCondition {
    /// A size container feature, implicitly parenthesized.
    Feature(ContainerSizeFeature),
    /// A negation of a condition.
    Not(Box<ContainerCondition>),
    /// A set of joint operations.
    Operation {
        /// The operator for the conditions.
        operator: Operator,
        /// The conditions for the operator.
        // PERF(port): was ArrayListUnmanaged fed input.arena() (parser arena);
        // Phase B decides bun_alloc::ArenaVec<'bump, _> vs global Vec crate-wide.
        conditions: Vec<ContainerCondition>,
    },
    /// A style query.
    Style(StyleQuery),
}

impl ToCss for ContainerCondition {
    fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        self.condition_to_css(dest)
    }
}

impl QueryCondition for ContainerCondition {
    type Feature = ContainerSizeFeature;

    fn as_feature(&self) -> Option<&ContainerSizeFeature> {
        if let Self::Feature(f) = self {
            Some(f)
        } else {
            None
        }
    }
    fn as_not(&self) -> Option<&Self> {
        if let Self::Not(c) = self {
            Some(c)
        } else {
            None
        }
    }
    fn as_operation(&self) -> Option<(Operator, &[Self])> {
        if let Self::Operation {
            operator,
            conditions,
        } = self
        {
            Some((*operator, conditions))
        } else {
            None
        }
    }
    fn feature_to_css(
        f: &ContainerSizeFeature,
        dest: &mut Printer,
    ) -> core::result::Result<(), PrintErr> {
        f.to_css(dest)
    }
    fn extra_to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        let Self::Style(query) = self else {
            unreachable!()
        };
        dest.write_str("style(")?;
        query.to_css(dest)?;
        dest.write_char(b')')
    }

    fn parse_feature(input: &mut css::Parser) -> css::Result<Self> {
        let feature = QueryFeature::<ContainerSizeFeatureId>::parse(input)?;
        Ok(ContainerCondition::Feature(feature))
    }
    fn create_negation(condition: Box<Self>) -> Self {
        ContainerCondition::Not(condition)
    }
    fn create_operation(operator: Operator, conditions: Vec<Self>) -> Self {
        ContainerCondition::Operation {
            operator,
            conditions,
        }
    }
    fn parse_style_query(input: &mut css::Parser) -> css::Result<Self> {
        use crate::media_query::QueryConditionFlags;
        // Zig defined a local `Fns` struct with two callbacks; in Rust pass closures.
        input.parse_nested_block(|i| {
            if let Ok(res) = i.try_parse(|i2| {
                media_query::parse_query_condition::<StyleQuery>(i2, QueryConditionFlags::ALLOW_OR)
            }) {
                return Ok(ContainerCondition::Style(res));
            }
            Ok(ContainerCondition::Style(StyleQuery::parse_feature(i)?))
        })
    }
    fn needs_parens(&self, parent_operator: Option<Operator>, targets: &css::Targets) -> bool {
        match self {
            ContainerCondition::Not(_) => true,
            ContainerCondition::Operation { operator, .. } => Some(*operator) == parent_operator,
            ContainerCondition::Feature(f) => f.needs_parens(parent_operator, targets),
            ContainerCondition::Style(_) => false,
        }
    }
}

impl ContainerCondition {
    pub fn deep_clone(&self, bump: &bun_alloc::Arena) -> Self {
        // PORT NOTE: `css.implementDeepClone` variant-walk. `QueryFeature<F>`
        // routes through `dc::query_feature` (Clone is faithful — see note
        // there); `Operator` is `Copy`.
        match self {
            Self::Feature(f) => Self::Feature(super::dc::query_feature(f, bump)),
            Self::Not(c) => Self::Not(Box::new(c.deep_clone(bump))),
            Self::Operation {
                operator,
                conditions,
            } => Self::Operation {
                operator: *operator,
                conditions: conditions.iter().map(|c| c.deep_clone(bump)).collect(),
            },
            Self::Style(q) => Self::Style(q.deep_clone(bump)),
        }
    }
}

// ─── ContainerCondition parse ─────────────────────────────────────────────
impl ContainerCondition {
    pub fn parse(input: &mut css::Parser) -> css::Result<ContainerCondition> {
        use crate::media_query::QueryConditionFlags;
        media_query::parse_query_condition::<ContainerCondition>(
            input,
            QueryConditionFlags::ALLOW_OR | QueryConditionFlags::ALLOW_STYLE,
        )
    }
}

/// A [@container](https://drafts.csswg.org/css-contain-3/#container-rule) rule.
pub struct ContainerRule<R> {
    /// The name of the container.
    pub name: Option<ContainerName>,
    /// The container condition.
    pub condition: ContainerCondition,
    /// The rules within the `@container` rule.
    pub rules: CssRuleList<R>,
    /// The location of the rule in the source file.
    pub loc: Location,
}

impl<R> ContainerRule<R> {
    pub fn to_css(&self, dest: &mut Printer) -> core::result::Result<(), PrintErr> {
        // #[cfg(feature = "sourcemap")]
        // dest.add_mapping(self.loc);

        dest.write_str("@container ")?;
        if let Some(name) = &self.name {
            name.to_css(dest)?;
            dest.write_char(b' ')?;
        }

        // Don't downlevel range syntax in container queries.
        let exclude = dest.targets.exclude;
        // Zig: bun.bits.insert(css.targets.Features, &dest.targets.exclude, .media_queries);
        dest.targets.exclude.insert(css::Features::MEDIA_QUERIES);
        self.condition.to_css(dest)?;
        dest.targets.exclude = exclude;

        dest.block(|d| {
            d.newline()?;
            self.rules.to_css(d)
        })
    }
}

impl<R> ContainerRule<R> {
    pub fn deep_clone<'bump>(&self, bump: &'bump bun_alloc::Arena) -> Self
    where
        R: css::generics::DeepClone<'bump>,
    {
        // PORT NOTE: `css.implementDeepClone` field-walk.
        Self {
            name: self.name.as_ref().map(|n| n.deep_clone(bump)),
            condition: self.condition.deep_clone(bump),
            rules: self.rules.deep_clone(bump),
            loc: self.loc,
        }
    }
}

// ported from: src/css/rules/container.zig
